keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubelet currently creates some IPTables chains at startup. In the past
+it actually used some of them, but with [the removal of dockershim] it
+no longer does. Additionally, for historical reasons, it creates some
+IPTables chains that are duplicates of chains also created by
+kube-proxy, and some that are only used by kube-proxy, but which
+kube-proxy requires kubelet to create on its behalf. (The initial
+comment in [kubernetes #82125] has more details on the history of how
+we got to where we are now, or at least to where we were before
+dockershim was removed.)
+
+We should clean this up so that kubelet no longer creates any IPTables
+chains, and that kube-proxy creates all of the IPTables chains it
+needs.
+
+[the removal of dockershim]: ../../sig-node/2221-remove-dockershim
+[kubernetes #82125]: https://github.com/kubernetes/kubernetes/issues/82125
+
+## Motivation
+
+### Goals
+
+- Rewrite kube-proxy's packet processing to no longer use the
+  `KUBE-MARK-DROP` chain created by kubelet.
+
+    - Ensure that the proxy implementations in kpng get updated
+      accordingly as well.
+
+- Remove all (or possibly just "most"?) IPTables chain management from
+  kubelet.
+
+    - Remove the `--iptables-drop-bit` and `--iptables-masquerade-bit`
+      arguments from kubelet.
+
+    - Ensure that users and third-party software that make assumptions
+      about kubelet's IPTables rules have time to get updated for the
+      change in rule creation.
+
+- Document for the future that Kubernetes's IPTables chains are meant
+  for its internal use only, do not constitute API, and should not be
+  relied on by external components.
+
+### Non-Goals
+
+- Changing `KUBE-MARK-MASQ` and `KUBE-MARK-DROP` to use the connection
+  mark or connection label rather than the packet label (as discussed
+  in [kubernetes #78948]).
+
+[kubernetes #78948]: https://github.com/kubernetes/kubernetes/issues/78948
+
+```
+<<[UNRESOLVED other-kubelet-network-cruft ]>>
+
+Is there any other related legacy networking cruft in kubelet that it
+would make sense to remove/move at the same time as the IPTables
+stuff?
```

Comment:
```
$ git grep -i NonMasq
cmd/kubelet/app/options/options.go:     // nonMasqueradeCIDR configures masquerading: traffic to IPs outside this range will use IP masquerade.
cmd/kubelet/app/options/options.go:     NonMasqueradeCIDR string
cmd/kubelet/app/options/options.go:             NonMasqueradeCIDR:       "10.0.0.0/8",
cmd/kubelet/app/options/options.go:     fs.StringVar(&f.NonMasqueradeCIDR, "non-masquerade-cidr", f.NonMasqueradeCIDR, "Traffic to IPs outside this range will use IP masquerade. Set to '0.0.0.0/0' to never masquerade.")
cmd/kubelet/app/server.go:              s.NonMasqueradeCIDR)
pkg/kubelet/kubelet.go: nonMasqueradeCIDR string) error {
```

Reply:
`git pull`
:slightly_smiling_face: 
(removed in #107096)